### PR TITLE
feat: re-adds `Self` typing with Python 3.8+ support

### PR DESCRIPTION
**Pull request by cofin**
[//]: # "By submitting this pull request, you agree to:"
[//]: # "- follow [Jolt's Code of Conduct](https://github.com/jolt-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)"
[//]: # "- follow [Jolt's contribution guidelines](https://github.com/jolt-org/.github/blob/main/CONTRIBUTING.md)"

### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description
[//]: # "Please describe your pull request for new release changelog purposes"

- 

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

-

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ dependencies = [
   "alembic>=1.12.0",
   "typing-extensions>=4.0.0",
   "greenlet; sys_platform == \"darwin\"",
+  "typing-extensions; python_version < \"3.11\""
```
Comment (peterschutt) [inline pyproject.toml:32]: IDT this actually constrains anything b/c its also declared as a dependency on line 30.

```
(.venv) peter@pop-os:~/PycharmProjects/aa-312-test$ python --version
Python 3.12.2
(.venv) peter@pop-os:~/PycharmProjects/aa-312-test$ python -m pip freeze
(.venv) peter@pop-os:~/PycharmProjects/aa-312-test$ python -m pip install git+https://github.com/jolt-org/advanced-alchemy.git@self-typing
Collecting git+https://github.com/jolt-org/advanced-alchemy.git@self-typing
--snipped--
(.venv) peter@pop-os:~/PycharmProjects/aa-312-test$ python -m pip freeze
advanced_alchemy @ git+https://github.com/jolt-org/advanced-alchemy.git@ee7bd61a0901420f603faf9b5d4cd7f1eae6fdd1
alembic==1.13.1
greenlet==3.0.3
Mako==1.3.2
MarkupSafe==2.1.5
SQLAlchemy==2.0.29
typing_extensions==4.10.0
(.venv) peter@pop-os:~/PycharmProjects/aa-312-test$ 
```

File: advanced_alchemy/operations.py [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
+import sys
 from typing import TYPE_CHECKING, Any
 
 from sqlalchemy import ClauseElement, ColumnElement, UpdateBase
 from sqlalchemy.ext.compiler import compiles
 
+if sys.version_info < (3, 11):
+    from typing_extensions import Self
+else:
+    from typing import Self
```
Comment (peterschutt) [inline advanced_alchemy/operations.py:12]: And so given that `typing_extensions` is already in your environment, you could import this directly from there in all cases and let them handle this conditional. E.g, their source for `Self` is [this](https://github.com/python/typing_extensions/blob/94bec447d6f7b9d3625ef0e688a0b0f9e487e951/src/typing_extensions.py#L1976-L1994):

```py


if hasattr(typing, "Self"):  # 3.11+
    Self = typing.Self
else:
    @_SpecialForm
    def Self(self, params):
        """Used to spell the type of "self" in classes.

        Example::

          from typing import Self

          class ReturnsSelf:
              def parse(self, data: bytes) -> Self:
                  ...
                  return self

        """

        raise TypeError(f"{self} is not subscriptable")
```

File: pyproject.toml (hunk 1 above)
Comment (cofin) [reply pyproject.toml:32]: Oops.  I meant to actually remove the other one :)

Comment (cofin) [reply pyproject.toml:32]: And actually, I'm going to just close this one.  I see that we are using `typing_extensions` in other places as well.

**Pull request closed without merging**